File: apps/app/src/hooks/market-vision/bbwp.ts
/* eslint-disable no-console */
'use client'

import type { OHLCVDataPoint, SeriesDataPoint } from './market-vision-config'
import { ema, hullMA, rma, sma, stdev, vwma, wma } from './technical-indicators'

export type BBWPPriceSource = 'close' | 'open' | 'high' | 'low' | 'hlc3'
export type BBWPMovingAverageType = 'SMA' | 'EMA' | 'WMA' | 'RMA' | 'HMA' | 'VWMA'

export interface BBWPConfig {
  priceSource: BBWPPriceSource
  basisType: BBWPMovingAverageType
  basisLength: number
  lookback: number

  maType: BBWPMovingAverageType
  maLength: number

  extremeHigh: number
  extremeLow: number
}

export interface BBWPResult {
  bbwp: SeriesDataPoint[]
  ma: SeriesDataPoint[]
  levels: {
    high: SeriesDataPoint[]
    mid: SeriesDataPoint[]
    low: SeriesDataPoint[]
    extremeHigh: SeriesDataPoint[]
    extremeLow: SeriesDataPoint[]
  }
}

export const DEFAULT_BBWP_CONFIG: BBWPConfig = {
  priceSource: 'close',
  basisType: 'SMA',
  basisLength: 7,
  lookback: 100,

  maType: 'SMA',
  maLength: 5,

  extremeHigh: 98,
  extremeLow: 2,
}

function clampNumber(value: number, min: number, max: number): number {
  if (!Number.isFinite(value)) return min
  return Math.min(max, Math.max(min, value))
}

function getPriceSeries(data: OHLCVDataPoint[], source: BBWPPriceSource): number[] {
  switch (source) {
    case 'open':
      return data.map((d) => d.open)
    case 'high':
      return data.map((d) => d.high)
    case 'low':
      return data.map((d) => d.low)
    case 'hlc3':
      return data.map((d) => (d.high + d.low + d.close) / 3)
    default:
      return data.map((d) => d.close)
  }
}

function basisMa(type: BBWPMovingAverageType, values: number[], len: number, volumes: number[]): number[] {
  switch (type) {
    case 'SMA':
      return sma(values, len)
    case 'EMA':
      return ema(values, len)
    case 'WMA':
      return wma(values, len)
    case 'RMA':
      return rma(values, len)
    case 'HMA':
      return hullMA(values, len)
    case 'VWMA':
      return vwma(values, volumes, len)
    default:
      return sma(values, len)
  }
}

function binarySearchRightmostLikePine(sortedAsc: number[], value: number): number {
  // Mirrors Pine Script's `array.binary_search_rightmost()` behavior used in the provided script:
  // - If value is found, return the index of the rightmost matching element.
  // - If not found, return the insertion index to the right (first index where element > value).
  //
  // This is implemented via upper-bound + equality check.
  let lo = 0
  let hi = sortedAsc.length

  while (lo < hi) {
    const mid = (lo + hi) >> 1
    const midValue = sortedAsc[mid]
    if (midValue != null && midValue <= value) lo = mid + 1
    else hi = mid
  }

  const upperBound = lo // first index with element > value
  const rightmostEqualIdx = upperBound - 1
  if (rightmostEqualIdx >= 0 && sortedAsc[rightmostEqualIdx] === value) return rightmostEqualIdx
  return upperBound
}

function rollingPercentileRankRightmost(
  values: Array<number | null>,
  bbwLen: number,
  lookback: number
): Array<number | null> {
  const rawValues: number[] = []
  const sortedValues: number[] = []
  const out: Array<number | null> = new Array(values.length).fill(null)

  for (let i = 0; i < values.length; i++) {
    const v = values[i]

    // Match Pine condition: `if bar_index >= _bbwLen`
    if (i < bbwLen) continue
    if (v == null || !Number.isFinite(v)) continue

    const idxOrInsert = binarySearchRightmostLikePine(sortedValues, v)
    const denom = sortedValues.length
    out[i] = denom > 0 ? (idxOrInsert * 100.0) / denom : null

    rawValues.push(v)
    sortedValues.splice(idxOrInsert, 0, v)

    if (rawValues.length > lookback) {
      const oldValue = rawValues.shift()
      if (oldValue != null) {
        const removalIndex = binarySearchRightmostLikePine(sortedValues, oldValue)
        // Pine expects this removal index to be in-bounds for found values.
        if (removalIndex >= 0 && removalIndex < sortedValues.length) sortedValues.splice(removalIndex, 1)
      }
    }
  }

  return out
}

function movingAverageNullable(
  type: BBWPMovingAverageType,
  values: Array<number | null>,
  len: number,
  volumes: number[]
): Array<number | null> {
  if (len <= 0) return new Array(values.length).fill(null)

  // For BBWP MA overlay, we mimic Pine behavior by propagating nulls until enough valid samples exist.
  // We compute MA on a dense array while tracking validity.
  const dense: number[] = values.map((v) => (v == null || !Number.isFinite(v) ? Number.NaN : v))
  const computed = basisMa(type, dense.map((v) => (Number.isFinite(v) ? v : 0)), len, volumes)

  const out: Array<number | null> = new Array(values.length).fill(null)
  let validCount = 0
  for (let i = 0; i < values.length; i++) {
    const v = values[i]
    if (v != null && Number.isFinite(v)) validCount++

    // Require at least `len` valid values in the prefix window.
    // This is a pragmatic match for Pine’s NA propagation on early samples.
    if (validCount < len) continue

    const c = computed[i]
    if (c != null && Number.isFinite(c)) out[i] = c
  }
  return out
}

export function calculateBBWP(data: OHLCVDataPoint[], config: Partial<BBWPConfig> = {}): BBWPResult {
  const finalConfig: BBWPConfig = { ...DEFAULT_BBWP_CONFIG, ...config }
  const basisLength = Math.max(1, Math.floor(finalConfig.basisLength))
  const lookback = Math.max(10, Math.floor(finalConfig.lookback))
  const maLength = Math.max(1, Math.floor(finalConfig.maLength))

  const extremeHigh = clampNumber(finalConfig.extremeHigh, 50, 100)
  const extremeLow = clampNumber(finalConfig.extremeLow, 0, 50)

  if (!data || data.length === 0) {
    return {
      bbwp: [],
      ma: [],
      levels: { high: [], mid: [], low: [], extremeHigh: [], extremeLow: [] },
    }
  }

  const times = data.map((d) => d.time)
  const volumes = data.map((d) => (Number.isFinite(d.volume) ? d.volume : 0))
  const price = getPriceSeries(data, finalConfig.priceSource)

  const basis = basisMa(finalConfig.basisType, price, basisLength, volumes)
  const deviation = stdev(price, basisLength)

  const bbw: Array<number | null> = new Array(data.length).fill(null)
  for (let i = 0; i < data.length; i++) {
    const b = basis[i] ?? Number.NaN
    const dev = deviation[i] ?? Number.NaN
    if (!Number.isFinite(b) || !Number.isFinite(dev) || b === 0) continue
    bbw[i] = (2 * dev) / b
  }

  const bbwpValues = rollingPercentileRankRightmost(bbw, basisLength, lookback).map((v) =>
    v == null ? null : clampNumber(v, 0, 100)
  )

  const maValues = movingAverageNullable(finalConfig.maType, bbwpValues, maLength, volumes).map((v) =>
    v == null ? null : clampNumber(v, 0, 100)
  )

  const bbwpSeries: SeriesDataPoint[] = []
  const maSeries: SeriesDataPoint[] = []

  const levelHigh: SeriesDataPoint[] = []
  const levelMid: SeriesDataPoint[] = []
  const levelLow: SeriesDataPoint[] = []
  const levelExtremeHigh: SeriesDataPoint[] = []
  const levelExtremeLow: SeriesDataPoint[] = []

  for (let i = 0; i < times.length; i++) {
    const time = times[i]
    if (time == null) continue

    levelHigh.push({ time, value: 100 })
    levelMid.push({ time, value: 50 })
    levelLow.push({ time, value: 0 })
    levelExtremeHigh.push({ time, value: extremeHigh })
    levelExtremeLow.push({ time, value: extremeLow })

    const v = bbwpValues[i]
    if (v != null && Number.isFinite(v)) bbwpSeries.push({ time, value: v })

    const m = maValues[i]
    if (m != null && Number.isFinite(m)) maSeries.push({ time, value: m })
  }

  return {
    bbwp: bbwpSeries,
    ma: maSeries,
    levels: {
      high: levelHigh,
      mid: levelMid,
      low: levelLow,
      extremeHigh: levelExtremeHigh,
      extremeLow: levelExtremeLow,
    },
  }
}

File: apps/ios/Packages/AggrCore/Sources/AggrCore/Indicators/BBWP.swift
import Foundation

/// Port of `bbwp.ts` — Bollinger Band Width Percentile.
public enum BBWP {
  public enum MAType: String, Sendable, Codable { case SMA, EMA, WMA, RMA, HMA, VWMA }
  public enum PriceSource: String, Sendable, Codable { case close, open, high, low, hlc3 }

  public struct Config: Sendable, Hashable {
    public var priceSource: PriceSource = .close
    public var basisType: MAType = .SMA
    public var basisLength = 7
    public var lookback = 100
    public var maType: MAType = .SMA
    public var maLength = 5
    public var extremeHigh = 98.0
    public var extremeLow = 2.0
    public init() {}
    public static let `default` = Config()
  }

  public struct Result: Sendable, Hashable {
    public var bbwp: [TimePoint]
    public var ma: [TimePoint]
    public var extremeHigh: Double
    public var extremeLow: Double
    public static let empty = Result(bbwp: [], ma: [], extremeHigh: 98, extremeLow: 2)
    public var latest: Double? { bbwp.last?.value }
  }

  static func ma(_ type: MAType, _ values: [Double], _ len: Int, _ volumes: [Double]) -> [Double] {
    switch type {
    case .SMA: TechnicalIndicators.sma(values, len)
    case .EMA: TechnicalIndicators.ema(values, len)
    case .WMA: TechnicalIndicators.wma(values, len)
    case .RMA: TechnicalIndicators.rma(values, len)
    case .HMA: TechnicalIndicators.hullMA(values, len)
    case .VWMA: TechnicalIndicators.vwma(values, volumes, len)
    }
  }

  /// Pine `array.binary_search_rightmost` semantics.
  static func binarySearchRightmost(_ sorted: [Double], _ value: Double) -> Int {
    var lo = 0, hi = sorted.count
    while lo < hi { let mid = (lo + hi) >> 1; if sorted[mid] <= value { lo = mid + 1 } else { hi = mid } }
    let upper = lo
    let idx = upper - 1
    if idx >= 0, sorted[idx] == value { return idx }
    return upper
  }

  static func rollingPercentileRank(_ values: [Double?], bbwLen: Int, lookback: Int) -> [Double?] {
    var raw: [Double] = [], sorted: [Double] = []
    var out = [Double?](repeating: nil, count: values.count)
    for i in 0..<values.count {
      guard i >= bbwLen, let v = values[i], v.isFinite else { continue }
      let idx = binarySearchRightmost(sorted, v)
      out[i] = sorted.isEmpty ? nil : Double(idx) * 100 / Double(sorted.count)
      raw.append(v)
      sorted.insert(v, at: idx)
      if raw.count > lookback {
        let old = raw.removeFirst()
        let r = binarySearchRightmost(sorted, old)
        if r >= 0, r < sorted.count { sorted.remove(at: r) }
      }
    }
    return out
  }

  static func movingAverageNullable(_ type: MAType, _ values: [Double?], _ len: Int, _ volumes: [Double]) -> [Double?] {
    guard len > 0 else { return [Double?](repeating: nil, count: values.count) }
    let dense = values.map { ($0 ?? .nan).isFinite ? $0! : 0 }
    let computed = ma(type, dense, len, volumes)
    var out = [Double?](repeating: nil, count: values.count)
    var valid = 0
    for i in 0..<values.count {
      if let v = values[i], v.isFinite { valid += 1 }
      guard valid >= len, i < computed.count, computed[i].isFinite else { continue }
      out[i] = computed[i]
    }
    return out
  }

  public static func calculate(_ data: [OHLCVBar], config c: Config = .default) -> Result {
    guard !data.isEmpty else { return .empty }
    let basisLength = max(1, c.basisLength), lookback = max(10, c.lookback), maLength = max(1, c.maLength)
    let extremeHigh = min(100, max(50, c.extremeHigh)), extremeLow = min(50, max(0, c.extremeLow))
    let volumes = data.map { $0.volume.isFinite ? $0.volume : 0 }
    let price: [Double] = switch c.priceSource {
    case .open: data.map(\.open); case .high: data.map(\.high); case .low: data.map(\.low); case .hlc3: data.map(\.hlc3); case .close: data.map(\.close)
    }
    let basis = ma(c.basisType, price, basisLength, volumes)
    let dev = TechnicalIndicators.stdev(price, basisLength)
    var bbw = [Double?](repeating: nil, count: data.count)
    for i in 0..<data.count {
      let b = basis[i], d = dev[i]
      guard b.isFinite, d.isFinite, b != 0 else { continue }
      bbw[i] = 2 * d / b
    }
    let bbwp = rollingPercentileRank(bbw, bbwLen: basisLength, lookback: lookback).map { $0.map { min(100, max(0, $0)) } }
    let maValues = movingAverageNullable(c.maType, bbwp, maLength, volumes).map { $0.map { min(100, max(0, $0)) } }
    var series: [TimePoint] = [], maSeries: [TimePoint] = []
    for i in 0..<data.count {
      if let v = bbwp[i], v.isFinite { series.append(TimePoint(epochSeconds: data[i].time, value: v)) }
      if let m = maValues[i], m.isFinite { maSeries.append(TimePoint(epochSeconds: data[i].time, value: m)) }
    }
    return Result(bbwp: series, ma: maSeries, extremeHigh: extremeHigh, extremeLow: extremeLow)
  }
}
